qbft/controller.go

```diff
@@ -0,0 +1,263 @@
+		if inst != nil {
+			if inst.GetHeight() == height {
+				return inst
+			}
+		}
+	}
+	return nil
+}
+
+// addNewInstance will add the new instance at index 0, pushing all other stored InstanceContainer one index up (ejecting last one if existing)
+func (i *InstanceContainer) addNewInstance(instance *Instance) {
+	for idx := HistoricalInstanceCapacity - 1; idx > 0; idx-- {
+		i[idx] = i[idx-1]
+	}
+	i[0] = instance
+}
+
+// Controller is a QBFT coordinator responsible for starting and following the entire life cycle of multiple QBFT InstanceContainer
+type Controller struct {
+	Identifier []byte
+	Height     Height // incremental Height for InstanceContainer
+	// StoredInstances stores the last HistoricalInstanceCapacity in an array for message processing purposes.
+	StoredInstances InstanceContainer
+	// FutureMsgsContainer holds all msgs from a higher height
+	FutureMsgsContainer map[types.OperatorID]Height // maps msg signer to height of higher height received msgs
+	Domain              types.DomainType
+	Share               *types.Share
+	config              IConfig
+}
+
+func NewController(
+	identifier []byte,
+	share *types.Share,
+	domain types.DomainType,
+	config IConfig,
+) *Controller {
+	return &Controller{
+		Identifier:          identifier,
+		Height:              -1, // as we bump the height when starting the first instance
+		Domain:              domain,
+		Share:               share,
+		StoredInstances:     InstanceContainer{},
+		FutureMsgsContainer: make(map[types.OperatorID]Height),
+		config:              config,
+	}
+}
+
+// StartNewInstance will start a new QBFT instance, if can't will return error
+func (c *Controller) StartNewInstance(value []byte) error {
+	if err := c.canStartInstance(c.Height+1, value); err != nil {
+		return errors.Wrap(err, "can't start new QBFT instance")
+	}
+
+	c.bumpHeight()
+	newInstance := c.addAndStoreNewInstance()
+	newInstance.Start(value, c.Height)
+
+	return nil
+}
+
+// ProcessMsg processes a new msg, returns decided message or error
+func (c *Controller) ProcessMsg(msg *SignedMessage) (*SignedMessage, error) {
+	if err := c.baseMsgValidation(msg); err != nil {
+		return nil, errors.Wrap(err, "invalid msg")
+	}
+
+	/**
+	Main controller processing flow
+	_______________________________
+	All decided msgs are processed the same, out of instance
+	All valid future msgs are saved in a container and can trigger highest decided futuremsg
+	All other msgs (not future or decided) are processed normally by an existing instance (if found)
+	*/
+	if isDecidedMsg(c.Share, msg) {
+		return c.UponDecided(msg)
+	} else if msg.Message.Height > c.Height {
+		return c.UponFutureMsg(msg)
+	} else {
+		return c.UponExistingInstanceMsg(msg)
+	}
+}
+
+func (c *Controller) UponExistingInstanceMsg(msg *SignedMessage) (*SignedMessage, error) {
+	inst := c.InstanceForHeight(msg.Message.Height)
+	if inst == nil {
+		return nil, errors.New("instance not found")
+	}
+
+	prevDecided, _ := inst.IsDecided()
```

Comment:
If this value turns out to be true, do you really need to continue processing the message?